File: app/fatigue_calculator/safte_model.py
# safte_model.py
# ---------------------------------------------------------------------
# Sleep, Activity, Fatigue, and Task Effectiveness (SAFTE) model
# Implements % Effectiveness with:
#   - Homeostatic reservoir (wake depletion, sleep recovery)
#   - Circadian drive (24 h + 12 h; phase can be offset for jet lag)
#   - Sleep inertia (exponential decay tied to "sleep intensity")
#   - Jet-lag: phase-shift events + optional first-order re-entrainment
#   - Sleep quality: per-episode efficiency scaling of recovery
#
# Notes:
# - Canonical SAFTE doesn't prescribe a re-entrainment ODE; we add an
#   optional exponential relaxation of internal-vs-local phase offset.
# - Sleep quality is operationalized as an efficiency (0..1) on recovery.
# ---------------------------------------------------------------------

from dataclasses import dataclass
from typing import List, Optional, Tuple, Dict
import math
import json

# --------------------------- Data types ------------------------------

@dataclass(frozen=True)
class SleepEpisode:
    """
    A single sleep episode in absolute hours.

    Args:
        start (float): start time (hours on a common absolute axis)
        end (float):   end time   (hours on the same axis), must be > start
        efficiency (float): sleep efficiency in [0, 1]; scales recovery
                            during this episode (sleep quality proxy).
    """
    start: float
    end: float
    efficiency: float = 1.0


@dataclass(frozen=True)
class PhaseShift:
    """
    An instantaneous local-time shift (e.g., crossing time zones).

    Args:
        time (float): absolute time (hours) of the shift
        delta_local_hours (float): change in local time (e.g., +6 for eastward)
        note (str): optional annotation
    """
    time: float
    delta_local_hours: float
    note: str = ""


# --------------------------- SAFTE model -----------------------------

class SAFTEModel:
    """
    SAFTE (Hursh et al., 2004) with pragmatic extensions:

    Outputs:
        % Effectiveness vs. time (None while sleeping)

    Core parameters (defaults reflect common SAFTE usage):
        R_c: reservoir capacity (units)                   -> 2880.0
        K:   wake depletion rate (units/min)              -> 0.5
        f:   sleep recovery rate constant (per min)       -> 0.01
        a_s: circadian effect on recovery (per min)       -> 0.05
        p:   24 h circadian peak phase (hours)            -> 18.0
        p_prime: 12 h harmonic phase (hours)              -> 3.0
        beta: 12 h harmonic relative amplitude            -> 0.5
        a1:  base circadian contribution to %             -> 7.0
        a2:  extra circadian contribution under fatigue % -> 5.0
        I_max: max sleep inertia drop (fraction)          -> 0.05 (5%)
        i_const: inertia decay constant                   -> 0.04
        inertia_window_min: inertia active minutes        -> 120

    Jet lag / re-entrainment:
        internal_minus_local (state): internal - local phase offset (h)
        reentrain_tau_h: time constant (hours) for exponential relaxation.
                         Set to None to disable dynamic re-entrainment.
                         Default: 24 h (≈1 h/day catch-up).

    Methods:
        circadian(t_local_h, internal_minus_local_h) -> C_t
        simulate(schedule, shifts, t0_h, t_end_h, dt_min) -> [(t, %Eff|None)]
        run_and_export(...) -> writes CSV (t_hour,effectiveness_percent)
        to_json_params() -> JSON of current parameters
    """

    # ---- Defaults (can be overridden via __init__(params=...)) ----
    R_c: float = 2880.0
    K: float = 0.5
    f: float = 0.01
    a_s: float = 0.05
    p: float = 18.0
    p_prime: float = 3.0
    beta: float = 0.5
    a1: float = 7.8
    a2: float = 5.0
    I_max: float = 0.05
    i_const: float = 0.04
    inertia_window_min: float = 120.0
    reentrain_tau_h: Optional[float] = 24.0

    # ---- Construction / overrides ----
    def __init__(self, params: Optional[Dict[str, float]] = None):
        if params:
            for k, v in params.items():
                if not hasattr(self, k):
                    raise ValueError(f"Unknown parameter: {k}")
                setattr(self, k, v)

    # ---------------- Circadian process (24 h + 12 h) ----------------

    def circadian(self, t_local_h: float, internal_minus_local_h: float) -> float:
        """
        Two-harmonic circadian drive C_t as a function of LOCAL time and
        internal-minus-local offset (jet-lag aware).

        Returns:
            C_t (dimensionless), roughly in [-1.5, +1.5] with default beta.
        """
        internal_phase = (t_local_h + internal_minus_local_h) % 24.0
        phase_rel = (internal_phase - self.p) % 24.0
        c1 = math.cos(2.0 * math.pi * phase_rel / 24.0)
        c2 = math.cos(4.0 * math.pi * (phase_rel - self.p_prime) / 24.0)
        return c1 + self.beta * c2

    # ------------------- Sleep inertia denominator -------------------

    def sleep_intensity_denominator(self, C_at_wake: float, R_at_wake: float) -> float:
        """
        Denominator in the inertia exponential; acts like 'sleep intensity'.
        """
        denom = self.a_s * C_at_wake + self.f * (self.R_c - R_at_wake)
        return max(denom, 1e-6)

    # ---------------------- Schedule validation ----------------------

    @staticmethod
    def _validate_schedule(schedule: List[SleepEpisode]) -> None:
        if any(ep.end <= ep.start for ep in schedule):
            raise ValueError("Each SleepEpisode must have end > start.")
        # check overlaps
        s = sorted(schedule, key=lambda e: e.start)
        for i in range(1, len(s)):
            if s[i].start < s[i - 1].end:
                raise ValueError("Sleep episodes overlap; please merge or correct them.")
        if any(not (0.0 <= ep.efficiency <= 1.0) for ep in schedule):
            raise ValueError("SleepEpisode.efficiency must be within [0, 1].")

    # ------------------------ Simulation engine ----------------------

    def simulate(
        self,
        schedule: List[SleepEpisode],
        shifts: List[PhaseShift],
        t0_h: float,
        t_end_h: float,
        dt_min: int = 1,
    ) -> List[Tuple[float, Optional[float]]]:
        """
        Simulate % Effectiveness across [t0_h, t_end_h] in steps of dt_min.

        Args:
            schedule: list of SleepEpisode (absolute hours)
            shifts:   list of PhaseShift (absolute hours)
            t0_h:     start time (hours)
            t_end_h:  end time (hours)
            dt_min:   step (minutes); 1 min recommended for fidelity

        Returns:
            List of (t_local_hour, effectiveness_percent_or_None).
            When asleep, effectiveness is None.
        """
        if dt_min <= 0:
            raise ValueError("dt_min must be positive.")
        if t_end_h <= t0_h:
            raise ValueError("t_end_h must be > t0_h.")
        self._validate_schedule(schedule)

        schedule = sorted(schedule, key=lambda e: e.start)
        shifts = sorted(shifts, key=lambda e: e.time)

        # Internal minus local offset (captures jet lag); 0 at start
        internal_minus_local = 0.0
        shift_idx = 0

        # Find active sleep episode at time t
        def active_sleep(t: float) -> Optional[SleepEpisode]:
            # schedules are typically small; linear scan suffices
            for ep in schedule:
                if ep.start <= t < ep.end:
                    return ep
            return None

        t = t0_h
        dt_h = dt_min / 60.0

        # Homeostatic reservoir and inertia bookkeeping
        R = self.R_c
        asleep = active_sleep(t) is not None
        last_wake_time: Optional[float] = None
        R_at_last_wake: float = R
        C_at_last_wake: float = 0.0

        if not asleep:
            last_wake_time = t
            C_at_last_wake = self.circadian(t, internal_minus_local)
            R_at_last_wake = R

        results: List[Tuple[float, Optional[float]]] = []

        # --------- Main loop ---------
        while t <= t_end_h + 1e-9:
            # Apply instantaneous phase shifts at this time (if any)
            while shift_idx < len(shifts) and abs(t - shifts[shift_idx].time) < 1e-9:
                # Local time jumps by +delta; internal clock unchanged:
                # internal_minus_local := internal_minus_local - delta_local_hours
                internal_minus_local -= shifts[shift_idx].delta_local_hours
                shift_idx += 1

            # Optional re-entrainment (exponential relaxation toward 0 offset)
            if self.reentrain_tau_h is not None and self.reentrain_tau_h > 0:
                internal_minus_local -= (internal_minus_local / self.reentrain_tau_h) * dt_h

            # State transitions
            ep = active_sleep(t)
            if asleep:
                if ep is None:
                    # Waking up now
                    asleep = False
                    last_wake_time = t
                    C_at_last_wake = self.circadian(t, internal_minus_local)
                    R_at_last_wake = R
            else:
                if ep is not None:
                    # Falling asleep now
                    asleep = True

            # Update reservoir
            C_val = self.circadian(t, internal_minus_local)
            if asleep:
                eff = ep.efficiency if ep is not None else 1.0
                # Exponential-like recovery minus circadian term
                dR_dt = (self.f * eff) * (self.R_c - R) - self.a_s * C_val
                R = min(self.R_c, max(0.0, R + dR_dt * dt_min))
            else:
                # Linear depletion while awake
                R = max(0.0, R - self.K * dt_min)

            # Compute % Effectiveness during wake and sleep.
            # During sleep, inertia is zero but we still report the latent
            # effectiveness based on reservoir and circadian state so the UI
            # can display continuous predictions.
            # Compute in percent space (classic SAFTE style) to avoid
            # saturation artifacts and to align with validated behavior.
            homeo_percent = 100.0 * (R / self.R_c)
            circ_percent = (self.a1 + self.a2 * (1.0 - (R / self.R_c))) * C_val
            inertia_frac = 0.0
            if not asleep and last_wake_time is not None:
                mins_since_wake = (t - last_wake_time) * 60.0
                if 0.0 <= mins_since_wake <= self.inertia_window_min:
                    denom = self.sleep_intensity_denominator(
                        C_at_wake=C_at_last_wake, R_at_wake=R_at_last_wake
                    )
                    inertia_frac = - self.I_max * math.exp(
                        - self.i_const * mins_since_wake / denom
                    )
            inertia_percent = 100.0 * inertia_frac
            # Normalization to map to percent effectiveness
            norm = 96.7
            E_percent = 100.0 * (homeo_percent + circ_percent + inertia_percent) / norm
            # Physiological bounding per literature: [~30, 100]
            E_percent = max(30.0, min(100.0, E_percent))
            results.append((t, E_percent))

            t += dt_h

        return results

    # ---------------- Convenience / export / introspection ------------

    def run_and_export(
        self,
        schedule: List[SleepEpisode],
        shifts: List[PhaseShift],
        t0_h: float,
        t_end_h: float,
        dt_min: int = 5,
        out_csv_path: Optional[str] = None,
    ) -> List[Tuple[float, Optional[float]]]:
        """
        Simulate and optionally export to CSV with headers:
            t_hour,effectiveness_percent
        """
        data = self.simulate(schedule, shifts, t0_h, t_end_h, dt_min)
        if out_csv_path:
            with open(out_csv_path, "w", encoding="utf-8") as f:
                f.write("t_hour,effectiveness_percent\n")
                for t, e in data:
                    f.write(f"{t:.4f},{'' if e is None else f'{e:.4f}'}\n")
        return data

    def to_json_params(self) -> str:
        """Return current parameters as pretty JSON."""
        keys = [
            "R_c", "K", "f", "a_s", "p", "p_prime", "beta",
            "a1", "a2", "I_max", "i_const",
            "inertia_window_min", "reentrain_tau_h",
        ]
        return json.dumps({k: getattr(self, k) for k in keys}, indent=2)


# ------------------------------ Demo ---------------------------------

if __name__ == "__main__":
    # Example: three nights of sleep, with an eastward 6 h phase shift before night 3.
    model = SAFTEModel(params={"reentrain_tau_h": 24.0})  # ~1 h/day re-entrainment

    schedule = [
        SleepEpisode(23.0, 31.0, efficiency=0.95),  # Night 1: 23:00–07:00, slightly fragmented
        SleepEpisode(47.0, 55.0, efficiency=1.00),  # Night 2
        SleepEpisode(71.0, 79.0, efficiency=0.85),  # Night 3 (after time-zone jump)
    ]
    shifts = [
        PhaseShift(time=70.0, delta_local_hours=+6, note="Eastward flight at t=70 h"),
    ]

    results = model.run_and_export(
        schedule=schedule,
        shifts=shifts,
        t0_h=0.0,
        t_end_h=96.0,
        dt_min=15,
        out_csv_path=None  # set a path (e.g., "safte_demo.csv") to export
    )

    # Print a few samples
    for t, eff in results[::8][:10]:
        status = "asleep" if eff is None else f"{eff:6.2f}%"
        print(f"t={t:6.2f} h, %Effectiveness={status}")

File: frontend/src/lib/safte-model.ts
// Author: Dr Diego Malpica MD
// ---------------------------------------------------------------------------
// Reservoir-Based SAFTE Model  (Hursh et al. 2004 / DRDC Peng & Bouak 2015)
// Shared module used by both Research and Operational tabs.
//
// Equations (matching backend safte_model.py):
//   E(t) = 100 * (homeo% + circ% + inertia%) / norm
//   homeo% = 100 * R_t / R_c
//   circ%  = (a1 + a2 * (1 - R_t/R_c)) * C_t
//   C_t    = cos(2π(t-p)/24) + β·cos(4π(t-p')/24)
//   R_t    = R_{t-1} - K·Δt          (wake)
//   R_t    = R_{t-1} + f·(R_c-R)·Δt  (sleep, simplified)
// ---------------------------------------------------------------------------

/** SAFTE parameters — identical to safte_model.py defaults */
export const SAFTE = {
  R_c: 2880.0,        // Reservoir capacity (units)
  K: 0.5,             // Wake depletion rate (units / min)
  f: 0.01,            // Sleep recovery rate constant (per min)
  a_s: 0.05,          // Circadian effect on recovery (per min)
  p: 18.0,            // 24 h circadian peak phase (hours)
  p_prime: 3.0,       // 12 h harmonic phase offset (hours)
  beta: 0.5,          // 12 h harmonic relative amplitude
  a1: 7.8,            // Base circadian contribution (%)
  a2: 5.0,            // Fatigue-dependent circadian (%)
  norm: 96.7,         // Normalization constant (%)
} as const;

export interface SAFTEForecast {
  hours: number[];
  effectiveness: number[];
  reservoir_pct: number[];   // R/R_c as percentage (Process S)
  circadian_drive: number[]; // Raw C_t (Process C)
  clockHours: number[];      // Absolute clock hours for each point
  isAsleep: boolean[];       // Whether subject is asleep at each point
}

/** Two-harmonic circadian drive — matches safte_model.py circadian() */
export function circadianDrive(clockHour: number): number {
  const phaseRel = (clockHour - SAFTE.p + 24) % 24;
  const c1 = Math.cos((2 * Math.PI * phaseRel) / 24);
  const c2 = Math.cos((4 * Math.PI * (phaseRel - SAFTE.p_prime)) / 24);
  return c1 + SAFTE.beta * c2;
}

/**
 * Derive initial reservoir level from effectiveness.
 * Inverts: E = 100 * (100*R/R_c + (a1 + a2*(1-R/R_c))*C_t) / norm
 */
export function reservoirFromEffectiveness(E_pct: number, C_t: number): number {
  const numerator = (E_pct * SAFTE.norm) / 100 - (SAFTE.a1 + SAFTE.a2) * C_t;
  const denominator = 100 - SAFTE.a2 * C_t;
  if (Math.abs(denominator) < 1e-6) return 0.9;
  const ratio = numerator / denominator;
  return Math.max(0, Math.min(1, ratio));
}

/**
 * Compute instantaneous effectiveness from reservoir and clock hour.
 * Used by operational tab for single-point queries.
 */
export function computeEffectiveness(R: number, clockHour: number): number {
  const C_t = circadianDrive(clockHour);
  const homeo_pct = 100 * (R / SAFTE.R_c);
  const circ_pct = (SAFTE.a1 + SAFTE.a2 * (1 - R / SAFTE.R_c)) * C_t;
  const E_pct = (100 * (homeo_pct + circ_pct)) / SAFTE.norm;
  return Math.max(30, Math.min(100, Math.round(E_pct * 10) / 10));
}

/**
 * Default sleep schedule: sleep from bedtime to bedtime + sleepDuration each night.
 * Returns a function that checks if a given clock hour is within a sleep period.
 */
function defaultSleepCheck(
  bedtimeHour: number,
  sleepDurationH: number,
): (clockHour: number) => boolean {
  return (clockHour: number) => {
    // Normalize to bedtime-relative hours
    const rel = (clockHour - bedtimeHour + 24) % 24;
    return rel < sleepDurationH;
  };
}

/**
 * Multi-day SAFTE forecast generator.
 *
 * @param baseEffectiveness  Current effectiveness % from backend
 * @param sleepDebtHours     Current accumulated sleep debt (hours)
 * @param predictionDays     Number of days to forecast (1–7)
 * @param bedtimeHour        Usual bedtime in 24 h format (default 23)
 * @param sleepDurationH     Usual sleep duration in hours (default 7)
 */
export function generateSAFTEForecast(
  baseEffectiveness: number,
  sleepDebtHours: number,
  predictionDays: number = 1,
  bedtimeHour: number = 23,
  sleepDurationH: number = 7,
): SAFTEForecast {
  const now = new Date();
  const currentHour = now.getHours() + now.getMinutes() / 60;
  const dt_min = 30;
  const numPoints = predictionDays * 48; // 48 per day (30-min steps)

  // Initialize reservoir from backend effectiveness
  const C_now = circadianDrive(currentHour);
  let R = reservoirFromEffectiveness(baseEffectiveness, C_now) * SAFTE.R_c;

  // If sleep debt is significant, ensure reservoir reflects it
  const debtDepletion = SAFTE.K * 60 * Math.max(0, sleepDebtHours);
  const R_from_debt = SAFTE.R_c - debtDepletion;
  R = Math.min(R, Math.max(0, R_from_debt));

  const isAsleepFn = defaultSleepCheck(bedtimeHour, sleepDurationH);

  const hours: number[] = [];
  const effectiveness: number[] = [];
  const reservoir_pct: number[] = [];
  const circadian_values: number[] = [];
  const clockHours: number[] = [];
  const isAsleep: boolean[] = [];

  for (let i = 0; i < numPoints; i++) {
    const h = i * 0.5;
    hours.push(h);

    const clockHour = (currentHour + h) % 24;
    clockHours.push(clockHour);
    const C_t = circadianDrive(clockHour);
    const sleeping = isAsleepFn(clockHour);
    isAsleep.push(sleeping);

    // --- Reservoir dynamics ---
    if (i > 0) {
      if (sleeping) {
        // Exponential recovery during sleep (matching safte_model.py line 247)
        const dR = SAFTE.f * (SAFTE.R_c - R) * dt_min - SAFTE.a_s * C_t * dt_min;
        R = Math.min(SAFTE.R_c, Math.max(0, R + dR));
      } else {
        // Linear depletion during wakefulness
        R = Math.max(0, R - SAFTE.K * dt_min);
      }
    }

    // --- Effectiveness (safte_model.py lines 259-276) ---
    const homeo_pct = 100 * (R / SAFTE.R_c);
    const circ_pct = (SAFTE.a1 + SAFTE.a2 * (1 - R / SAFTE.R_c)) * C_t;
    const E_pct = (100 * (homeo_pct + circ_pct)) / SAFTE.norm;

    effectiveness.push(Math.max(30, Math.min(100, Math.round(E_pct * 10) / 10)));
    reservoir_pct.push(Math.round((R / SAFTE.R_c) * 1000) / 10);
    circadian_values.push(Math.round(C_t * 100) / 100);
  }

  return { hours, effectiveness, reservoir_pct, circadian_drive: circadian_values, clockHours, isAsleep };
}

/** Quick current-effectiveness for operational tab (avoids full simulation) */
export function currentSAFTEEffectiveness(sleepDebtHours: number = 0): number {
  const now = new Date();
  const clockHour = now.getHours() + now.getMinutes() / 60;
  const debtDepletion = SAFTE.K * 60 * Math.max(0, sleepDebtHours);
  const R = Math.max(0, SAFTE.R_c - debtDepletion);
  return computeEffectiveness(R, clockHour);
}
